### PR TITLE
Warnings for adjoint diff with finite shots

### PR DESCRIPTION
**Pull request by albi3ro**
Fixes Issue #1092 

When a QNode is created with a device with finite shots and adjoint differentiation, the User receives a warning reminding that adjoint differentiation is always exact.

When the adjoint jacobian method is called on a device with finite shots, the User also receives a warning to the same effect.

I only warn instead of interrupting with an error, as finite shots doesn't actually break the code.

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -577,6 +577,8 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 
 <h3>Bug fixes</h3>
 
+* Warns when adjoint differentiation specified or called on device with finite shots.
```
Comment (albi3ro) [inline .github/CHANGELOG.md:580]:
```suggestion
* Warns when adjoint differentiation specified or called on a device with finite shots.
[(#1406)](https://github.com/PennyLaneAI/pennylane/pull/1406)
```

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -854,6 +854,13 @@ def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
             if not hasattr(m.obs, "base_name"):
                 m.obs.base_name = None  # This is needed for when the observable is a tensor product
 
+        if self.shots is not None:
+            warnings.warn(
+                "Adjoint diff called on device with finite shots. Adjoint diff always"
```
Comment (antalszava) [inline pennylane/_qubit_device.py:859]: Would be great to rephrase the message a bit such that the term "adjoint differentiation" aligns well in there (e.g., "Requested the gradient to be computed with finite shots, this computation is always exact when using the adjoint differentiation method." or so).

File: tests/test_qubit_device_adjoint_jacobian.py [hunk 3]
```diff
@@ -39,6 +39,17 @@ def test_not_expval(self, dev):
         with pytest.raises(qml.QuantumFunctionError, match="Adjoint differentiation method does"):
             dev.adjoint_jacobian(tape)
 
+    def test_finite_shots_warns(self):
```
Comment (antalszava) [inline tests/test_qubit_device_adjoint_jacobian.py:42]: It would be nice to have a test with the QNode UI too. Should be quite similar, as this one is already a great case! :)

File: pennylane/qnode.py [hunk 4]
```diff
@@ -419,6 +419,13 @@ def _validate_adjoint_method(device, interface):
                 f"The {device.short_name} device does not support adjoint differentiation."
             )
 
+        if device.shots is not None:
+            warnings.warn(
+                "Adjoint diff requested on device with finite shots. Adjoint"
```
Comment (antalszava) [inline pennylane/qnode.py:424]: How come there's a warning here too? :thinking: Would it suffice to have it only on the device level?

On another note, a similar comment related to the message would be great to rephrase.

**Review by antalszava — COMMENTED** [review]
Hi @albi3ro, thank you for looking into this! :slightly_smiling_face: I think having a warning is an equally nice solution overall.

> if diff_method="best" and analytic=False, modify the QNode logic to reject adjoint/reversible and instead use parameter-shift. This way gradients will be estimated with a finite number of shots.

This doesn't seem to be addressed in this PR, any reasons?

Also, how come that the `reversible` diff method was not included?

**Comment by josh146** [comment]
Thanks for tackling this @albi3ro!

> if diff_method="best" and analytic=False, modify the QNode logic to reject adjoint/reversible and instead use parameter-shift. This way gradients will be estimated with a finite number of shots.

I think it’s best to leave this for now. I’ve been thinking about this a lot, and I wonder if the best approach is to instead, for next release, modify the adjoint method to *sample* from the final inner products/expectation values, rather than computing them directly.

This will be a nice, straightforward way of having the adjoint method support finite-shots, while avoiding the need to drop back down to parameter-shift.

**Comment by albi3ro** [comment]
@antalszava  Thanks for reminding me about reversible ;) I forgot...

I don't think we should switch the differentiation method to parameter shift behind the scenes, as parameter shift can send off many expensive circuit executions that the user might not be expecting.

I also included the warnings in two places:
* I think we need to warn when the QNode is initialized before anything happens, because that's the best place for users to do something about it
* We also need to warn in the actual `adjoint_jacobian` function, because shots may have changed since the QNode was constructed or users could access the method via a different route.

File: pennylane/qnode.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/qnode.py:424]: This is when users can most easily do something about the warning.  They haven't already invested in a forward pass.

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/_qubit_device.py:859]: I'll think about this a bit more.

**Comment by antalszava** [comment]
> I don't think we should switch the differentiation method to parameter shift behind the scenes, as parameter shift can send off many expensive circuit executions that the user might not be expecting.

Originally the following was proposed:

> if diff_method="best" and analytic=False, modify the QNode logic to reject adjoint/reversible and instead use parameter-shift. This way gradients will be estimated with a finite number of shots.

That is, the user would specify `diff_method="best"` and `shots=1000` (some finite shots). If the user is aware of the meaning of these options then it would seem sensible to use the parameter shift rule as that's the best diff method for finite shots. There is of course the possibility that these options are specified without knowing the exact details, though not sure if we should tackle it by falling back to exact and adjoint or even a warning. :thinking:


> I think we need to warn when the QNode is initialized before anything happens, because that's the best place for users to do something about it

Would be nice to have a test case for this: the QNode is initialized, but not evaluated and the user sees a warning.

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/_qubit_device.py:859]: Or I'll just use what you said ;)

**Comment by albi3ro** [comment]
@josh146  I don't it makes sense for adjoint diff or reversible to ever be computed with finite shots.  Finite shots is used to mimic the behaviour of actual quantum devices, but adjoint diff and reversible aren't computations you can reasonably make on actual hardware.

Unless we know the distributions will match that of parameter shift?

**Comment by albi3ro** [comment]
Also, neither adjoint nor reversible will ever get selected if `diff_method="best"`.  We should change that, but it should be a different PR.

File: tests/tape/test_qnode.py [hunk 5]
```diff
@@ -266,9 +266,49 @@ def test_validate_adjoint_finite_shots(self):
...
         dev = qml.device("default.qubit", wires=1, shots=1)
 
-        with pytest.warns(UserWarning, match="Adjoint diff requested on device with finite shots."):
+        with pytest.warns(
+            UserWarning, match="Requested adjoint differentiation to be computed with finite shots."
+        ):
             QNode._validate_adjoint_method(dev, "autograd")
 
+    def test_adjoint_finite_shots(self):
+        """Tests that UserWarning is raised on QNode construction when device has finite shots"""
```
Comment (antalszava) [inline tests/tape/test_qnode.py:275]:
```suggestion
        """Tests that UserWarning is raised with the adjoint differentiation method
        on QNode construction when the device has finite shots
        """
```

File: tests/tape/test_qnode.py [hunk 6]
```diff
@@ -266,9 +266,49 @@ def test_validate_adjoint_finite_shots(self):
...
+        with pytest.warns(
+            UserWarning, match="Requested adjoint differentiation to be computed with finite shots."
+        ):
+
+            @qml.qnode(dev, diff_method="adjoint")
+            def circ():
+                return qml.expval(qml.PauliZ(0))
+
+    def test_validate_reversible_finite_shots(self):
+        """Test that a UserWarning is raised when device has finite shots"""
```
Comment (antalszava) [inline tests/tape/test_qnode.py:288]:
```suggestion
        """Test that a UserWarning is raised when validating the reversible differentiation method
        and using a device that has finite shots
        """
```

File: tests/tape/test_qnode.py [hunk 7]
```diff
@@ -266,9 +266,49 @@ def test_validate_adjoint_finite_shots(self):
...
+        dev = qml.device("default.qubit", wires=1, shots=1)
+
+        with pytest.warns(
+            UserWarning,
+            match="Requested reversible differentiation to be computed with finite shots.",
+        ):
+            QNode._validate_reversible_method(dev, "autograd")
+
+    def test_reversible_finite_shots(self):
+        """Tests that UserWarning is raised on QNode construction when device has finite shots"""
```
Comment (antalszava) [inline tests/tape/test_qnode.py:299]:
```suggestion
        """Tests that UserWarning is raised with the reversible differentiation method
        on QNode construction when the device has finite shots
        """
```

File: tests/tape/test_reversible.py [hunk 8]
```diff
@@ -160,6 +160,21 @@ def test_phaseshift_exception(self):
 class TestGradients:
     """Jacobian integration tests for qubit expectations."""
 
+    def test_finite_shots_warning(self):
+        """Test warning raised when calling jacobian with a device with finite shots"""
```
Comment (antalszava) [inline tests/tape/test_reversible.py:164]:
```suggestion
        """Test that a warning is raised when calling the jacobian with a device with finite shots"""
```

File: tests/tape/test_reversible.py [hunk 9]
```diff
@@ -369,6 +384,27 @@ def test_gradient_gate_with_multiple_parameters(self, tol):
 class TestQNodeIntegration:
     """Test QNode integration with the reversible method"""
 
+    def test_finite_shots_warning(self):
+        """ "Tests that a warning is raised if used on a device with finite shots"""
```
Comment (antalszava) [inline tests/tape/test_reversible.py:388]:
```suggestion
        """Test that a warning is raised when calling the jacobian with a device with finite shots"""
```

**Review by antalszava — APPROVED** [review]
Looks good to me! :+1: Thank you! :tada:

**Comment by josh146** [comment]
> I don't it makes sense for adjoint diff or reversible to ever be computed with finite shots.

I think the big use-case in my mind: a user is simulating a VQA with a gradient-based optimizer in analytic mode, and wants to swap over to finite-number of shots to see how the optimization is affected.

Allowing adjoint to support finite-shots by sampling from the inner-products used to compute the adjoint expectation values would allow this without resulting in p^2 scaling as required by the parameter shift rule.

> Unless we know the distributions will match that of parameter shift?

This is a good question, and worth investigating!
